pecos/core/xmc/inference.hpp

```diff
@@ -212,90 +212,83 @@ namespace pecos {
 
-        // Maps a matrix row index into an index of the row_ptr array below
-        unordered_map<index_type, index_type> row_hash;
         index_type col_begin; // The column this chunk starts at (inclusive)
         index_type col_end; // The column this chunk ends at (exclusive)
-        mem_index_type* row_ptr; // An array of where rows begin in hash_chunked_matrix_t::entries
-        bool b_has_explicit_bias; // Whether or not this chunk has an explicit bias term
+        index_type nnz_rows; // The number of non-zero rows in this chunk
+        // Using index_type for struct padding
+        index_type b_has_explicit_bias; // Whether or not this chunk has an explicit bias term
 
-        hash_chunk_t() :
-            row_ptr(nullptr),
-            b_has_explicit_bias(false) {
-        }
+        // This struct does not hold memory, all pointers are only views
+        unordered_map<index_type, index_type> row_hash; // Maps a matrix row index into an index of the row_ptr array below
+        mem_index_type* row_ptr; // An array of where rows begin in hash_chunked_matrix_t::entries
 
-        ~hash_chunk_t() {
-            if (row_ptr) {
-                delete[] row_ptr;
-            }
-        }
 
-        void set_empty() {
-            row_ptr = nullptr;
+        hash_chunk_view_t() :
+            col_begin(0),
+            col_end(0),
+            nnz_rows(0),
+            b_has_explicit_bias(false),
+            row_ptr(nullptr) {
         }
 
-        void init(const index_type nnz_rows) {
-            row_ptr = new mem_index_type[nnz_rows + 1];
+        void set_row(index_type row, index_type i_row, mem_index_type i_entry) {
+            row_ptr[i_row] = i_entry;
+            row_hash[row] = i_row;
         }
 
-        void set_row(index_type row, index_type row_indx, mem_index_type ptr) {
-            row_ptr[row_indx] = ptr;
-            row_hash[row] = row_indx;
+        index_type row_ptr_size() {
+            return nnz_rows == 0 ? (0) : (nnz_rows + 1);
         }
     };
 
-    struct bin_search_chunk_t {
+    // View of a binary search chunk, does not hold any memory
+    struct bin_search_chunk_view_t {
         typedef typename csc_t::index_type index_type;
         typedef typename csc_t::mem_index_type mem_index_type;
         typedef typename chunk_entry_t::value_type value_type;
 
         index_type col_begin; // The column this chunk starts at (inclusive)
         index_type col_end; // The column this chunk ends at (exclusive)
-        index_type* row_idx; // Stores the row id of each nnz row
-        mem_index_type* row_ptr; // Stores the pointer to data of each nnz row
-        index_type nnz_rows;
-        bool b_has_explicit_bias; // Whether or not this chunk has an explicit bias term
+        index_type nnz_rows; // The number of non-zero rows in this chunk
+        // Using index_type for struct padding
+        index_type b_has_explicit_bias; // Whether or not this chunk has an explicit bias term, 0=false
```

Comment:
given that this variable is no longer a boolean, we might want to consider to remove the "b_" prefix.